Vpn.Service/Manager.cs

```diff
@@ -79,24 +80,46 @@ private async Task HandleClientMessageStart(ReplyableRpcMessage<ServiceMessage,
     {
         try
         {
-            // TODO: if the credentials and URL are identical and the server
-            //       version hasn't changed we should not do anything
-            // TODO: this should be broken out into it's own method
-            _logger.LogInformation("ClientMessage.Start: testing server '{ServerUrl}'", message.Message.Start.CoderUrl);
-            var client = new CoderApiClient(message.Message.Start.CoderUrl, message.Message.Start.ApiToken);
-            var buildInfo = await client.GetBuildInfo(ct);
-            _logger.LogInformation("ClientMessage.Start: server version '{ServerVersion}'", buildInfo.Version);
-            var serverVersion = SemVersion.Parse(buildInfo.Version);
-            if (!serverVersion.Satisfies(ServerVersionRange))
-                throw new InvalidOperationException(
-                    $"Server version '{serverVersion}' is not within required server version range '{ServerVersionRange}'");
-            var user = await client.GetUser(User.Me, ct);
-            _logger.LogInformation("ClientMessage.Start: authenticated as '{Username}'", user.Username);
+            var serverVersion =
+                await CheckServerVersionAndCredentials(message.Message.Start.CoderUrl, message.Message.Start.ApiToken,
+                    ct);
+            if (_tunnelSupervisor.IsRunning && _lastStartRequest != null &&
+                _lastStartRequest.Equals(message.Message.Start) && _lastServerVersion == serverVersion)
+            {
+                // The client is requesting to start an identical tunnel while
+                // we're already running it.
+                _logger.LogInformation("ClientMessage.Start: Ignoring duplicate start request");
+                await message.SendReply(new ServiceMessage
+                {
+                    Start = new StartResponse
+                    {
+                        Success = true,
+                    },
+                }, ct);
+                return;
+            }
 
+            _lastStartRequest = message.Message.Start;
+            _lastServerVersion = serverVersion;
+
+            // Stop the tunnel if it's running so we don't have to worry about
+            // permissions issues when replacing the binary.
+            await _tunnelSupervisor.StopAsync(ct);
             await DownloadTunnelBinaryAsync(message.Message.Start.CoderUrl, serverVersion, ct);
-            await _tunnelSupervisor.StartAsync(_config.TunnelBinaryPath, HandleTunnelRpcMessage,
-                HandleTunnelRpcError,
+            await _tunnelSupervisor.StartAsync(_config.TunnelBinaryPath, HandleTunnelRpcMessage, HandleTunnelRpcError,
```

Comment:
same with this section: we need to ensure that stops and starts are serialized so that you don't get `stop, stop, start, start` if two clients connect around the same time.

Reply:
Added a semaphore that fixes both of these issues